### PR TITLE
Refactor widgets and layer metadata

### LINKED ISSUE #55: Set a reasonable upper bound of points that we want to render
napari currently can't render more than X points. Find a reasonable value for X (probably something between 100K and 1M) and change the code to subset to this number of points (load only those in-memory from dask and show only them).

### LINKED ISSUE #87: Add element to layer metadata
On main, currently the layer does not know which SpatialElement it is visualizing. This makes it impossible to change visibility of the individual layers when a coordinate system is selected in which the spatial element is not present.

### PR DESCRIPTION
**Pull request by melonora**
This PR moves the widgets in interactive to a script specific for data widgets. On top of that it merges the code from @LucaMarconato regarding adding metadata to the layers with the current code on main. 

Utility functions have been moved to utils. Performance has been improved by submitting points and polygons data based on the maximum amount to be visualized, before performing any other calculations.

Also a bug has been fixed for the point data (x and y coordinates were not swapped, napari expects yx).

Closes #87 and #66, potentially closes #55

### CODE REVIEW

**Comment by melonora** [comment]
@rahulbshrestha @tothmarcella Hi both, Could you please have a look at the tests that are failing and the code in general? It seems to be due to a missing attribute and a mismatch in dimensions in anndata.obs and blobs.shape. 

I think the changes are following what is intended so it could be that we have to change some tests.

**Comment by rahulbshrestha** [comment]
Hi @melonora! Yup, I'll take a look.

**Comment by melonora** [comment]
For the regionprops, we could also use dask regionprops. Would there be any objection to that?

File: src/napari_spatialdata/_utils.py [hunk 1]
```diff
@@ -221,3 +203,272 @@ def _points_inside_triangles(points: NDArrayA, triangles: NDArrayA) -> NDArrayA:
...
+    # merged.loc[index_of_background, "v"] = 0
+    merged["v"] = merged["v"].astype(int)
+
+    assert len(annotating_rows) == len(merged)
+    assert annotating_rows.obs[instance_key].tolist() == merged["v"].tolist()
+
+    merged_centroids = merged[["cx", "cy"]].to_numpy()
+    assert len(merged_centroids) == len(merged)
+    annotating_rows.obsm["spatial"] = merged_centroids
+    annotating_rows.obsm["region_radius"] = np.array([10.0] * len(merged_centroids))  # arbitrary value
```
Comment (melonora) [inline src/napari_spatialdata/_utils.py:339]: This part is due to squidpy legacy code. Can this also be removed?

File: src/napari_spatialdata/_utils.py [hunk 2]
```diff
@@ -221,3 +203,272 @@ def _points_inside_triangles(points: NDArrayA, triangles: NDArrayA) -> NDArrayA:
...
+            assert len(v) == 1
+            xdata = v.__iter__().__next__()
+            list_of_xdata.append(xdata)
+        new_raster = list_of_xdata
+
+    return new_raster, rgb
+
+
+def _init_colors_for_obs(adata: AnnData) -> AnnData:
+    from scanpy.plotting._utils import _set_colors_for_categorical_obs
```
Comment (giovp) [inline src/napari_spatialdata/_utils.py:263]: this we copied over it frm scanpy so to not be dependent on scanpy anymore, it's in `categorical_utils.py`

File: setup.cfg [hunk 3]
```diff
@@ -45,6 +45,7 @@ install_requires =
     napari-matplotlib
     scikit-learn
     spatialdata
+    scanpy
```
Comment (giovp) [inline setup.cfg:48]: remove scanpy  as requirement

File: src/napari_spatialdata/_utils.py [hunk 4]
```diff
@@ -221,3 +203,272 @@ def _points_inside_triangles(points: NDArrayA, triangles: NDArrayA) -> NDArrayA:
...
+
+
+def _get_mapping_info(annotation_table: AnnData) -> tuple[str, str, str]:
+    regions = annotation_table.uns["spatialdata_attrs"]["region"]
+    regions_key = annotation_table.uns["spatialdata_attrs"]["region_key"]
+    instance_key = annotation_table.uns["spatialdata_attrs"]["instance_key"]
+    return regions, regions_key, instance_key
+
+
+def _find_annotation_for_labels(
```
Comment (giovp) [inline src/napari_spatialdata/_utils.py:290]: it's unclear why this is needed, in the implementation from main there shouldn't be need for this. Is it possible that this is something you ported over from luca's branch and now it's here for legacy reason?

File: src/napari_spatialdata/_interactive.py [hunk 5]
```diff
@@ -1,203 +1,12 @@
...
-            self._sdata.labels[key],
-            name=key,
-            affine=affine,
-            metadata={
-                "adata": self._sdata.table[
-                    self._sdata.table.obs[self._sdata.table.uns["spatialdata_attrs"]["region_key"]] == key
-                ],
-                "labels_key": self._sdata.table.uns["spatialdata_attrs"]["instance_key"],
-            },
-        )
```
Comment (giovp) [inline src/napari_spatialdata/_interactive.py:167]: this part is replaced by the big/complex function below. I think this way to handle annotation metadata is much cleaner and would keep this.

File: src/napari_spatialdata/_utils.py [hunk 6]
```diff
@@ -221,3 +203,272 @@ def _points_inside_triangles(points: NDArrayA, triangles: NDArrayA) -> NDArrayA:
...
+    assert annotating_rows.obs[instance_key].tolist() == merged["v"].tolist()
+
+    merged_centroids = merged[["cx", "cy"]].to_numpy()
+    assert len(merged_centroids) == len(merged)
+    annotating_rows.obsm["spatial"] = merged_centroids
+    annotating_rows.obsm["region_radius"] = np.array([10.0] * len(merged_centroids))  # arbitrary value
+    return annotating_rows
+
+
+def _find_annotation_for_shapes(
```
Comment (giovp) [inline src/napari_spatialdata/_utils.py:343]: same here as before

File: src/napari_spatialdata/_utils.py [hunk 7]
```diff
@@ -221,3 +203,272 @@ def _points_inside_triangles(points: NDArrayA, triangles: NDArrayA) -> NDArrayA:
...
+    annotating_rows.uns[Key.uns.spatial][element_path] = {}
+    annotating_rows.uns[Key.uns.spatial][element_path][Key.uns.scalefactor_key] = {}
+    annotating_rows.uns[Key.uns.spatial][element_path][Key.uns.scalefactor_key]["tissue_hires_scalef"] = 1.0
+    annotating_rows.obsm["spatial"] = spatial
+    # workaround for the legacy code to support different sizes for different circles
+    annotating_rows.obsm["region_radius"] = radii
+    return annotating_rows
+
+
+def _find_annotation_for_regions(
```
Comment (giovp) [inline src/napari_spatialdata/_utils.py:399]: same here

File: src/napari_spatialdata/_utils.py [hunk 8]
```diff
@@ -221,3 +203,272 @@ def _points_inside_triangles(points: NDArrayA, triangles: NDArrayA) -> NDArrayA:
...
+            else:
+                raise TypeError(f"Unsupported shape type: {type(shape)}")
+        else:
+            raise ValueError(f"Unsupported element type: {type(base_element)}")
+        return table
+
+    return None
+
+
+def _get_ellipses_from_circles(centroids: NDArrayA, radii: NDArrayA) -> NDArrayA:
```
Comment (giovp) [inline src/napari_spatialdata/_utils.py:438]: same here

**Review by giovp — CHANGES_REQUESTED** [review]
I would revert back to what we had before regarding passing annotation metadata to layers and to add layers. imho the spatialdata widget should be kept the same as we had before. @tothmarcella @rahulbshrestha could you also give it a comment here thanks!

File: src/napari_spatialdata/_sdata_widgets.py [hunk 9]
```diff
@@ -0,0 +1,261 @@
...
+                f"being correct beyond a certain zoom level"
+            )
+            # TODO: If https://github.com/scverse/napari-spatialdata/issues/35 is fixed, use points for circles, faster
+            self._viewer.add_points(
+                coords_yx,
+                name=key,
+                size=2 * radii,
+                metadata=metadata,
+                affine=affine,
+            )
```
[the remark this answers is not in the document]
Comment (LucaMarconato) [reply src/napari_spatialdata/_sdata_widgets.py:129]: In `main` currently only ellipses are used, this is bad for performance and user experience, so definitely we should prefer always points to always shapes, I will make a PR. Re always points vs always shapes, I'll experiment a bit with small Visium datasets and get back to it.

[the remark this answers is not in the document]
Comment (melonora) [reply src/napari_spatialdata/_sdata_widgets.py:129]: Note that there is work ongoing in napari at the moment to actually fix this issue. Might already be included in 0.4.18 which is aimed for release next month

[the remark this answers is not in the document]
Comment (LucaMarconato) [reply src/napari_spatialdata/_sdata_widgets.py:129]: Very nice!

**Comment by melonora** [comment]
Closing this in favor of new PR

**Pull request closed without merging**